ext/ddtrace_profiling_native_extension/http_transport.c

```diff
@@ -0,0 +1,285 @@
+static void exporter_as_ruby_object_free(void *data);
+static VALUE exporter_to_ruby_object(ddprof_ffi_ProfileExporterV3* exporter);
+static VALUE _native_do_export(
+  VALUE self,
+  VALUE libddprof_exporter,
+  VALUE upload_timeout_milliseconds,
+  VALUE start_timespec_seconds,
+  VALUE start_timespec_nanoseconds,
+  VALUE finish_timespec_seconds,
+  VALUE finish_timespec_nanoseconds,
+  VALUE pprof_file_name,
+  VALUE pprof_data,
+  VALUE code_provenance_file_name,
+  VALUE code_provenance_data
+);
+static ddprof_ffi_Request *build_request(
+  ddprof_ffi_ProfileExporterV3 *exporter,
+  VALUE upload_timeout_milliseconds,
+  VALUE start_timespec_seconds,
+  VALUE start_timespec_nanoseconds,
+  VALUE finish_timespec_seconds,
+  VALUE finish_timespec_nanoseconds,
+  VALUE pprof_file_name,
+  VALUE pprof_data,
+  VALUE code_provenance_file_name,
+  VALUE code_provenance_data
+);
+static void *call_exporter_without_gvl(void *exporter_and_request);
+
+void http_transport_init(VALUE profiling_module) {
+  VALUE http_transport_class = rb_define_class_under(profiling_module, "HttpTransport", rb_cObject);
+
+  rb_define_singleton_method(
+    http_transport_class, "_native_create_agentless_exporter",  _native_create_agentless_exporter, 3
+  );
+  rb_define_singleton_method(
+    http_transport_class, "_native_create_agent_exporter",  _native_create_agent_exporter, 2
+  );
+  rb_define_singleton_method(
+    http_transport_class, "_native_do_export",  _native_do_export, 10
+  );
+
+  exporter_class = rb_define_class_under(http_transport_class, "Exporter", rb_cObject);
+  // This prevents creation of the exporter class outside of our extension, see https://bugs.ruby-lang.org/issues/18007
+  rb_undef_alloc_func(exporter_class);
+
+  ok_symbol = ID2SYM(rb_intern("ok"));
+  error_symbol = ID2SYM(rb_intern("error"));
+}
+
+inline static ddprof_ffi_ByteSlice byte_slice_from_ruby_string(VALUE string) {
+  Check_Type(string, T_STRING);
+  ddprof_ffi_ByteSlice byte_slice = {.ptr = (uint8_t *) StringValuePtr(string), .len = RSTRING_LEN(string)};
+  return byte_slice;
+}
+
+static VALUE _native_create_agentless_exporter(VALUE self, VALUE site, VALUE api_key, VALUE tags_as_array) {
+  Check_Type(site, T_STRING);
+  Check_Type(api_key, T_STRING);
+  Check_Type(tags_as_array, T_ARRAY);
+
+  return create_exporter(
+    ddprof_ffi_EndpointV3_agentless(
+      byte_slice_from_ruby_string(site),
+      byte_slice_from_ruby_string(api_key)
+    ),
+    tags_as_array
+  );
+}
+
+static VALUE _native_create_agent_exporter(VALUE self, VALUE base_url, VALUE tags_as_array) {
+  Check_Type(base_url, T_STRING);
+  Check_Type(tags_as_array, T_ARRAY);
+
+  return create_exporter(
+    ddprof_ffi_EndpointV3_agent(byte_slice_from_ruby_string(base_url)),
+    tags_as_array
+  );
+}
+
+static VALUE create_exporter(struct ddprof_ffi_EndpointV3 endpoint, VALUE tags_as_array) {
+  long tags_count = rb_array_len(tags_as_array);
+  ddprof_ffi_Tag converted_tags[tags_count];
```

Comment:
Does Ruby have macros/functions for optionally using `alloca` and falling back to `malloc` depending on the size and if the platform supports it?

PHP has [these macros](https://heap.space/xref/PHP-8.1/Zend/zend_portability.h?r=ee377744#348), and it's useful for cases like this where it _should_ be small enough to put on the stack but in case someone passes a large array you are protected against blowing your stack.

Reply:
That's... a great point, thanks! I dug around the documentation and the VM headers and there are a few things there (undocumented, of course!) but nothing that seemed a really good fit.

So in 65cbe6dd31fb3782f915ec7f2d4cf729b97ab45d I just changed this to a heap allocation. Shrug emoji :)